app/scripts/views/settings/recovery_codes.js

```diff
@@ -0,0 +1,151 @@
+
+const View = FormView.extend({
+  template: Template,
+  className: 'recovery-codes',
+  viewName: 'settings.two-step-authentication.recovery-codes',
+
+  events: {
+    'click .copy-codes': FormView.preventDefaultThen('_copyCodes'),
+    'click .download-codes': FormView.preventDefaultThen('_downloadCodes'),
+    'click .print-codes': FormView.preventDefaultThen('_printCodes'),
+    'click .replace-codes-link': FormView.preventDefaultThen('_replaceRecoveryCodes'),
+    'click .two-step-authentication-done': FormView.preventDefaultThen('_returnToTwoStepAuthentication')
+  },
+
+  _returnToTwoStepAuthentication() {
+    this.navigate('settings/two_step_authentication');
+  },
+
+  _copyCodes() {
+    // This copies the recovery codes to clipboard by creating a tiny transparent
+    // textArea with recovery code contents. Then it executes the
+    // browser `copy` command.
+    // Ref: https://stackoverflow.com/questions/400212/how-do-i-copy-to-the-clipboard-in-javascript
+    const textArea = this.window.document.createElement('textArea');
+    textArea.style.position = 'fixed';
+    textArea.style.top = 0;
+    textArea.style.left = 0;
+    textArea.style.width = '2em';
+    textArea.style.height = '2em';
+    textArea.style.padding = 0;
+    textArea.style.border = 'none';
+    textArea.style.outline = 'none';
+    textArea.style.boxShadow = 'none';
+    textArea.style.background = 'transparent';
+    textArea.value = this.recoveryCodesText;
+    this.window.document.body.appendChild(textArea);
+    textArea.focus();
+    textArea.select();
+    try {
+      this.window.document.execCommand('copy');
+      this._displaySuccess(t('Codes copied'));
+    } catch (err) {
+      this._displayError(t('Failed to copy codes'));
+    }
+    this.window.document.body.removeChild(textArea);
+  },
+
+  _downloadCodes() {
+    // This dynamically creates a link with a blob data of the recovery
+    // codes and clicks it to initiate download.
+    const codeBlob = new Blob([this.recoveryCodesText], {type: 'text/plain'});
+    const link = this.window.document.createElement('a');
+    link.download = 'Recovery Codes.txt';
+    link.href = URL.createObjectURL(codeBlob);
+    link.dataset.downloadurl = ['text/plain', link.download, link.href].join(':');
+    link.style.display = 'none';
+    this.window.document.body.appendChild(link);
+    link.click();
+    this.window.document.body.removeChild(link);
+  },
+
+  _printCodes() {
+    // We dynamically create a new window with recovery codes and attempt to
+    // print it.
+    let content = this.recoveryCodesText;
+    // Add line breaks for readablity
+    content = content.split('\n').join('</br>');
+    const printWindow = this.window.open('', 'Print', 'height=600,width=800');
+    printWindow.document.write('<html><head><title>Recovery Codes</title>');
+    printWindow.document.write('</head><body >');
+    printWindow.document.write(content);
+    printWindow.document.write('</body></html>');
```

Comment:
...oh, same question for `_downloadCodes` and `_printCodes`, is there a way we can render the markup from a template?

Both for correctness and developer happiness when editing. Working inside your favourite HTML editor, it can scream at you if/when you forget a closing tag or quote or whatever. Embedding it in the JS seems like a recipe for hard-to-spot little bugs.

Reply:
(although the correctness thing doesn't apply to the first two examples I realise, because they're all using proper DOM methods like `appendChild` or whatever. I dunno, maybe ignore me, I'm not very good at front-end stuff)

Reply:
Also, is there a potential future security footgun here of using `document.write` with unsanitised input? Admittedly I can't see a way for `content` to end up containing something malicious right now, but if `content` were the result of rendering a template at least we'd be depending on a layer that escaped e.g. `<script>` tags for us.

Reply:
> is there a potential future security footgun here

+1 to adding some belt-and-braces here; even if we can prove it's safe *now*, it's the sort of thing that could easily be overlooked in a future refactor.

Reply:
> but if content were the result of rendering a template

Updated this to use a mustache template which sanitizes the input. However, I couldn't find any way around using the `write` method. Open to suggestions if it is possible to create and open a new window.

Reply:
:+1: If it's being rendered from a template I think `write` is fine.